### PR TITLE
feat(linter): `no-array-constructor`

### LINKED ISSUE #67: feat(linter): implement `no-array-constructor`
This is going to be fun! 

We will be porting the `no-array-constructor` rule for JavaScript and TypeScript at the same time:

* https://eslint.org/docs/latest/rules/no-array-constructor
* https://typescript-eslint.io/rules/no-array-constructor

The goal of this task is to:

* iron out the details for implementing a rule
* draft a documentation for how to implement a rule
* nice to have - create a script to transform tests to our Rust, because it is going to be very time consuming to copy, paste and modify the tests

Your feedbacks will shape the future of this project.

I have also setup a [discord channel](https://discord.com/invite/9uXCAwqQZW) for quick communications.

Drop a comment if you wish to try this, so I can assign this issue to you.

### PR DESCRIPTION
**Pull request by amitdahan**
# Summary

Closes #67.

Wasn't actually sure about using `if-let-chains` as [predicted here](https://github.com/Boshen/oxc/issues/67#issuecomment-1445741198), would love to hear where that would fit.

### CODE REVIEW

File: crates/oxc_linter/src/rules/no_array_constructor.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            AstKind::NewExpression(NewExpression {
+                span,
+                arguments,
+                callee: Expression::Identifier(ident),
+                type_parameters,
+                ..
+            }) if arguments.len() != 1 && ident.name == "Array" && type_parameters.is_none() => {
+                ctx.diagnostic(NoArrayConstructorDiagnostic(*span));
+            }
+            _ => {}
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_array_constructor.rs:50]: Which style do you prefer?

```rust
        let (node, callee, arguments, type_parameters, optional) = match node.get().kind() {
            AstKind::CallExpression(call_expr) => (
                call_expr.node,
                &call_expr.callee,
                &call_expr.arguments,
                &call_expr.type_parameters,
                call_expr.optional,
            ),
            AstKind::NewExpression(new_expr) => (
                new_expr.node,
                &new_expr.callee,
                &new_expr.arguments,
                &new_expr.type_parameters,
                false,
            ),
            _ => return,
        };

        if let Expression::Identifier(ident) = callee
            && ident.name == "Array"
                && arguments.len() != 1
                && type_parameters.is_none()
                && !optional
        {
            ctx.diagnostic(NoArrayConstructorDiagnostic(*span));
            return

        }
```

File: crates/oxc_linter/src/rules/no_array_constructor.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-array-constructor): Disallow `Array` constructors")]
+#[diagnostic()]
```
Comment (Boshen) [inline crates/oxc_linter/src/rules/no_array_constructor.rs:14]: this needs a warning?

**Comment by amitdahan** [comment]
Also clippy was complaining about our inevitable enum naming 😅:

```
error: all variants have the same prefix: `No`
  --> crates/oxc_linter/src/rules.rs:20:1
   |
20 | / pub enum RuleEnum {
21 | |     NoDebugger(NoDebugger),
22 | |     NoEmpty(NoEmpty),
23 | |     NoArrayConstructor(NoArrayConstructor),
24 | | }
   | |_^
   |
   = help: remove the prefixes and use full paths to the variants instead of glob imports
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#enum_variant_names
   = note: `-D clippy::enum-variant-names` implied by `-D warnings`
```

**Comment by Boshen** [comment]
So overall what do you think?

* Is the overall experience good? 
* Do you feel more comfortable writing this in Rust compared to JavaScript / TypeScript?
* Will this be difficult for people who are super new to Rust?
* Are there anythings worth documenting?
* Are there things that should be improved?

**Comment by Boshen** [comment]
> Also clippy was complaining about our inevitable enum naming 😅:

Tell it to shut up lol. This boilerplate will be generated from a macro soon, so no worries.

File: crates/oxc_linter/src/rules/no_array_constructor.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_linter/src/rules/no_array_constructor.rs:50]: I don't really like the destructing pattern because it's hard to follow where the if-conditions are.

Comment (amitdahan) [reply crates/oxc_linter/src/rules/no_array_constructor.rs:50]: Definitely looks much better!

Is this using `if-let-chains`?

Is it here?
```rust
if let Expression::Identifier(ident) = callee
    && ident.name == "Array"
    && arguments.len() != 1
    && type_parameters.is_none()
    && !optional
{
    // ...
}
```

Comment (Boshen) [reply crates/oxc_linter/src/rules/no_array_constructor.rs:50]: yes this is, rustc will warn you if you remove the feature flag from `lib.rs`

**Comment by amitdahan** [comment]
> - Is the overall experience good?

IMO, yes! I don't even find the boilerplate to be all too bad, but once that's solved I'm sure it'll be even better!

> - Do you feel more comfortable writing this in Rust compared to JavaScript / TypeScript?

I don't have a lot of experience writing Rust, but yes, everything feels so correct & type-safe 😅 

> - Will this be difficult for people who are super new to Rust?

I'd say definitely not, as I identify as being super new to Rust and it was a good experience 😅 
After reading the original ESLint / TypeScript-ESLint flavor of this - it was pretty easy, but who knows what other rules have in store for us 😄 

> - Are there anythings worth documenting?

For now I felt the usual workflow of just running tests / using `cargo watch` and finally running `cargo insta` were all pretty self-documenting.

> - Are there things that should be improved?

I guess I'm not as well-versed to come up with improvement ideas just yet, so none so far.

---

> So overall what do you think?

Overall I'm very grateful for the opportunity for such well-defined contribution options, I'm a Rust beginner looking to learn and I'm very interested in the compiler space, so it fits the bill perfectly!

Lint is also a very chill way to start learning and contributing to OXC.

Also your guidance and mentorship seems unprecedented to me personally, very much appreciated @Boshen !

**Comment by Boshen** [comment]
Thank you! This is awesome!
